Allow arbitrary structure items in `let` expressions

nojb opened the pull request:
This PR contains the last bits needed to complete the plan exposed in #13835. Since #13839, the compiler is technically able to handle arbitrary structure items in `let` expressions. Here, we expose this possibility to the user by extending the language accepted by the parser.

After this PR, all kinds of structure items are allowed in `let` expressions, with the following two exceptions:

- `let` bindings (to avoid the seemingly strange syntax `let let x = 12 in ...`)
- `include` (as far as I can see, `let include` would be a less efficient equivalent to `let open`, and inviting confusion between the two).

The manual has been updated. I am planning to add some tests shortly.

nojb committed:
parser: allow arbitrary local structure items (except include, let bindings)

nojb committed:
Adapt manual

nojb committed:
Changes

v-gb commented on manual/src/refman/expr.etex:
```diff
@@ -1397,5 +1394,68 @@ equivalent to @module-path'.([|' expr '|])'@.
 
+\subsubsection*{sss:expr-struct-item}{Local structure items}
+\ikwd{let\@\texttt{type}}
+(Introduced in OCaml 5.5)
```
I think it'd be preferable to have this new section, but also remove the previous sections about local open/module/exception, and instead add a note saying "prior to ocaml to ocaml 5.5, only local open/module/exception were accepted".
This way, the manual describes the current syntax, instead of having the shape of the manual mirror the history of the syntax.

nojb replied:
Good point. I pushed in https://github.com/ocaml/ocaml/pull/14040/commits/df96e3207caee95f0b4583329e6162ddb97e3aab a rework of the manual additions. I didn't remove the "old" subsections because they contain remarks and clarifications which are useful for those special cases, but I put them under a general section explaining the general construct introduced in this PR.

See a preview over at https://nojb.github.io/ocaml-manual-wip/expr.html#ss%3Aexpr-local-items.

v-gb replied:
Sure, sounds good.

v-gb commented on manual/src/refman/expr.etex:
```diff
@@ -1397,5 +1394,68 @@ equivalent to @module-path'.([|' expr '|])'@.
+\end{caml_example}
+
+In a different direction, one can use this construct to toggle a warning
+thoughout the inner expression:
```
```suggestion
throughout the inner expression:
```

nojb replied:
Fixed, thanks.

v-gb reviewed (commented):
The change straightforwardly does what it says. I'm certainly in favor.

It'd be slightly better to have an example of syntax for each new local structure item, perhaps in testsuite/tests/exotic-syntax. Probably not strictly necessary, but it'd help make sure there's no "harmless" shift/reduce syntax conflict that prevents some construct from actually parsing, which I'm pretty sure I've seen happen before.

I suppose it's been discussed in past PRs, but I haven't seen such discussions, so feel free to ignore this: to me, the name Pexp_letitem seems more consistent with the naming convention of the AST. `Pexp_struct_item` would be expected  to have a payload  of `struct_item`.

Finally, just some things I wondering about while thinking about this change, you can skip.

In principle, `Pexp_seq (e1, e2)` could be replaced by `Pexp_struct_item (Pstr_eval e1, e2)` and `Pexp_let (bindings, e)` by `Pexp_struct_item (Pstr_value bindings, e)`, if the latter constructions weren't forbidden by an invariant. But things are good as they are, things can be refactor at any time if desired.

Although expressions are structure items, `let print_string "a" in ...` is not accepted (it wasn't clear from the description, and I was wondering since `let%foo print_string "a" in ..` would be better than `[%foo e1; e2]` or (yuck) `print_string "a" ;%foo e2`).

While trying things out in the toplevel, I tried to create a recursive module `let rec module M  = ... in`, which doesn't work. Instead you have to say `let module rec M ... in ...`. Not what I'd prefer to write, but obviously I understand why, and also who cares since recursive modules are very rare. Still, I wondered why we have these extra `let`: if we allowed `let f () = module rec M = ... in ()`, we'd get shift/reduce conflict on `(module Foo :`: is it a first class module or a local module? It's ambiguous at this point in the syntax. Now maybe it's possible to restructure the grammar to avoid this, but that's too much effort to try.

The similarity between `fun (type t) -> ..` and `let type t in ...` is a bit unfortunate, but oh well.

nojb commented:
Thanks @v-gb for the review! I will look at your comments in detail a bit later today, but here are some initial responses:

> It'd be slightly better to have an example of syntax for each new local structure item, perhaps in testsuite/tests/exotic-syntax. Probably not strictly necessary, but it'd help make sure there's no "harmless" shift/reduce syntax conflict that prevents some construct from actually parsing, which I'm pretty sure I've seen happen before.

Good idea. I will add this.

> I suppose it's been discussed in past PRs, but I haven't seen such discussions, so feel free to ignore this: to me, the name Pexp_letitem seems more consistent with the naming convention of the AST. `Pexp_struct_item` would be expected to have a payload of `struct_item`.

The precise name was chosen in #13839, but without much discussion. I was also told (offline) that the AST node name was a bit confusing. I am open to renaming it (since this patch has not been released, we can still do it). `Pexp_letitem` seems OK to me. @gasche: as having bikeshedded about the name of this constructor before; what do you think?

> Although expressions are structure items, `let print_string "a" in ...` is not accepted (it wasn't clear from the description, and I was wondering since `let%foo print_string "a" in ..` would be better than `[%foo e1; e2]` or (yuck) `print_string "a" ;%foo e2`).

Indeed, you are right this was not mentioned in the description. It is an interesting point. Let me think about it a bit more and get back here (I need to check there are no parsing issues if we were to allow it).

v-gb commented:
> > Although expressions are structure items, `let print_string "a" in ...` is not accepted (it wasn't clear from the description, and I was wondering since `let%foo print_string "a" in ..` would be better than `[%foo e1; e2]` or (yuck) `print_string "a" ;%foo e2`).
> 
> Indeed, you are right this was not mentioned in the description. It is an interesting point. Let me think about it a bit more and get back here (I need to check there are no parsing issues if we were to allow it).

Sorry, I wasn't suggesting to try this, I simply thought this was part of your change until I read the code. I think there's no chance that this can work. Both because accepting both `let expr in` and `let pattern = expr in` would conflict hard, and because accepting both `let+ a in` and `let a in` with unrelated meanings is weird.

nojb commented:
> I think there's no chance that this can work.

Yes, indeed, sorry, I have forgotten some of the details since I first wrote the code :)

nojb committed:
Update manual/src/refman/expr.etex

Co-authored-by: v-gb <valentin.gatienbaron@gmail.com>

nojb committed:
Add syntax test

nojb commented:
> It'd be slightly better to have an example of syntax for each new local structure item, perhaps in testsuite/tests/exotic-syntax. Probably not strictly necessary, but it'd help make sure there's no "harmless" shift/reduce syntax conflict that prevents some construct from actually parsing, which I'm pretty sure I've seen happen before.

Added in https://github.com/ocaml/ocaml/pull/14040/commits/8981bbcb47295781ce787a13cbe6b6e435ac8b27.

v-gb commented on testsuite/tests/parsetree/source.ml:
```diff
@@ -7525,3 +7525,26 @@ let là = function ça -> ça
+
+(* PR #14040 *)
+
+let () =
+  let [%%foo] [@@bar] in
+  let [@@@attr] in
+  let external foo: unit -> unit = "foo" [@@unboxed] in
+  let type 'a t = int in
+  let type 'a s in
+  let type 'a u = {a: int} [@@bar] in
```
Apparently code in this file is only parsed. Is there a particular reason you chose this? It feels like, if we have test code, we might as well ensure the typer and lambda translation don't choke.

nojb replied:
I will add semantic tests shortly. But extensions nodes (`[%%foo]`, etc) cannot easily be tested semantically, so I preferred to have separate syntax-only tests.

nojb replied:
I added some more tests in https://github.com/ocaml/ocaml/pull/14040/commits/5c17579bddf7fcc9dff309d6f7d4ff05dc6f6379

v-gb replied:
Sounds good. Can you mark this as done? I think I can't do it myself, or at least I looked for a way and couldn't find it.

nojb committed:
Add more tests

v-gb commented on testsuite/tests/exotic-syntax/exotic.ml:
```diff
@@ -159,3 +159,20 @@ function +1n -> ();;
+  let [@@@warning "-unused-var"] in
+  let v = (42, 12) in
+  assert (f v == v)
+  
```
Considering this feature is also tested in typing-misc/letitem.ml, I think it'd be net simpler to consolidate the tests by moving this section into letitem.ml (also, I'm the one who suggested adding tests here, seeing the rest of the file, they seem a bit out of place).

v-gb replied:
Not sure why github marked this as outdated, it still looks relevant to me.

nojb replied:
Moved the tests in 3889d2e72b2faa15addb6cc90072ac1a2f08e0c5, thanks.

314eter commented:
There are let expressions for classes and class types too.

```ocaml
class empty = let open M in object end
```

They only allow `let open`, and since they were not included in #14009 don't support extension points.

Should they be updated too, or are classes obscure enough to leave them untouched?

nojb commented:
> > I suppose it's been discussed in past PRs, but I haven't seen such discussions, so feel free to ignore this: to me, the name Pexp_letitem seems more consistent with the naming convention of the AST. `Pexp_struct_item` would be expected to have a payload of `struct_item`.
> 
> The precise name was chosen in #13839, but without much discussion. I was also told (offline) that the AST node name was a bit confusing. I am open to renaming it (since this patch has not been released, we can still do it). `Pexp_letitem` seems OK to me. 

I opened #14171 to discuss this point.

nojb committed:
check-typo

nojb commented:
> Should they be updated too, or are classes obscure enough to leave them untouched?

I wouldn't say that, but `let` bindings in class expressions have enough peculiarities (see eg #13178) that I think it would be better to treat them in a separate PR.

nojb committed:
Manual

v-gb commented on manual/src/refman/expr.etex:
```diff
@@ -1292,58 +1268,85 @@ otherwise the default is nonvariant. This is also the case for
 
-\subsection{ss:expr-other}{Other}
+\subsection{ss:expr-local-items}{Local structure items}
+(Introduced in OCaml 5.5)
```
```suggestion
(General case introduced in OCaml 5.5, only let open, non recursive let module and let exception were supported before)
```

(might need something around "let open" etc to mark them as keywords).

Otherwise it sounds like the whole section is new.

nojb replied:
Fixed in 6a5962f42474ce6fd6ec47d60cf7112bc89a08e4, thanks.

v-gb commented on manual/src/refman/expr.etex:
```diff
@@ -1292,58 +1268,85 @@ otherwise the default is nonvariant. This is also the case for
+\begin{center}
+@"let" "module" M = "struct" local-definition "let" body "=" expr "end" "in" M.body@
+\end{center}
+without the overhead of creating the actual module @M@ nor the binding @body@.
```
While this equivalence is correct, I was surprised when I read this, because this text is right after the definition of a let item, so it feels like it's saying "equivalently, we could have defined let items in terms of let module this way", which would not define how `let module` itself works.

Maybe we could say:
```
This construction has equivalent scoping and evaluation to this module expression:
`struct local-definition let body = expr end`
```

which is enough to give the intuition about "what does `let [@@@ocaml.warning "-4"] in ...` do?", without the recursivity.

nojb replied:
The problem is that `struct local-definition let body = expr end` evaluates to a structure, not the value of `expr`, so this could also be confusing.

The original explanation in terms of `let module` had been written when the special case of `let module` was explained in the manual before `let <struct item>`, but indeed that is no longer the case, so it doesn't make sense anymore.

nojb replied:
I went with your suggestion anyway, with a slight rewording: 04907dd48e679bad6fa4aa5da03102dc8b6afa2d

v-gb commented on manual/src/refman/expr.etex:
```diff
@@ -1292,58 +1268,85 @@ otherwise the default is nonvariant. This is also the case for
 
-\ikwd{assert\@\texttt{assert}}
+\subsubsection{sss:expr-let-type}{Local types}
+\ikwd{let\@\texttt{type}}
```
Minor, but maybe we should explain the very common items before the niche one, which local types will presumably be.
It feels like the order `let open`, `let module`, anything else might be most sensible ?

nojb replied:
Good point. Fixed in 3cf766be3ebe4ada345a974f8e0da052a4996243

v-gb commented on manual/src/refman/expr.etex:
```diff
@@ -1292,58 +1268,85 @@ otherwise the default is nonvariant. This is also the case for
-checking can be turned off with the "-noassert" compiler option.  In
-this case, @expr@ is not evaluated at all.
+The form @"let" typedef { "and" typedef } "in" expr@ introduces a type (or a
+collection of mutually recursive types) in the scope of an expression:
```
Shouldn't there be a an optional `rec`, if we mention recursive types ?

Well, even that is not enough, because further down, you talk about extensible type constructors, which have a different syntax entirely. In fact, it would probably make more sense to merge the extensible type constructors section into the `let exception` one somehow.

nojb replied:
> Shouldn't there be a an optional `rec`, if we mention recursive types ?

Fixed in f127dc095e5b519edebd8947be77cecf995fb7b8, thanks.

> Well, even that is not enough, because further down, you talk about extensible type constructors, which have a different syntax entirely. In fact, it would probably make more sense to merge the extensible type constructors section into the `let exception` one somehow.

Fixed in e452b3849fc86d27462a1b70f0e0facfddc1669d, thanks.

v-gb commented on manual/src/refman/expr.etex:
```diff
@@ -1292,58 +1268,85 @@ otherwise the default is nonvariant. This is also the case for
+\begin{caml_example}{toplevel}
+let gen () = let exception A in A
+
+let () = assert(gen () = gen ());;
```
Urh, it seems that github doesn't support commenting on unmodified parts of the code.

Anyway, the first sentence of the `let module` is now describing a syntax less general than is supported (in fact, the base description is already incorrect, as it claims that `let module F(A : B) = ... in ...` is not valid, which it is).

This should presumably refer to some kind of `module_decl` grammar item.

nojb replied:
There was no grammar item for module definitions, so I introduced one (`module-definition`) and used it here: 9cca85a293

Thanks.

v-gb commented on manual/src/refman/expr.etex:
```diff
@@ -1292,58 +1268,85 @@ otherwise the default is nonvariant. This is also the case for
 
-\subsubsection*{sss:expr-lazy}{Lazy expressions}
-\ikwd{lazy\@\texttt{lazy}}
+Extensible types can also be introduced in this way. The example below exhibits
```
Terminology issue: this is demonstrating the creation of an extensible type constructor, not extensible type.
The creation of an extensible type would be `let type t = .. in blabla`.

nojb replied:
Indeed, clarified the wording in 9598a0d17e

nojb committed:
Promote parser

gasche commented:
This PR needs an approval before being mergeable. @v-gb, out of curiosity, now that @nojb has taken your review comments into account, would you consider "approving" the PR? (That is, taking responsibility for saying that this is good in its current state and could be merged.)

nojb commented:
I am planning to go over the documentation aspects of this PR once more to make sure I took into account all of @v-gb's suggestions, because I have forgotten some of the details in the interim period. But another review pass and/or approval would be appreciated in any case :)

v-gb commented:
@gasche I do intend to approve, but not yet: there are still outstanding review comments.

gasche reviewed (approved):
I have reviewed the implementation part of this PR, which is merely a change to the parser now that most of the logic has been implement in #13839. It is fine. I am approving now, based on the idea that the tests and the manual will be in a good state thanks to @v-gb's review (once both of you have converged on potential manual changes).

nojb committed:
Merge remote-tracking branch 'upstream/trunk' into generalize_local_structure_items

nojb committed:
Revert Menhir's version bump

nojb committed:
manual: make it clear that some special cases already existed

nojb committed:
Try to explain the construct without using 'let module'

nojb committed:
Reorder special cases

nojb committed:
Move extensible types together with exceptions

nojb committed:
Add optional nonrec in local types explanation

nojb committed:
Clarify: types vs type constructors

nojb committed:
Grammar of 'let module' can be more general

nojb committed:
Move some tests to letitem.ml

nojb committed:
Add cross-reference

nojb commented:
@v-gb I think I addressed all your comments (hopefully I haven't missed any!). Can you take a look whenever convenient and let me know if you are content with the current state of the PR?

The updated manual can be previewed over at  https://nojb.github.io/ocaml-manual-wip/expr.html#ss%3Aexpr-local-items.

Thanks again for your thorough review!

v-gb reviewed (approved):
Looks good to me, modulo the PR I made against this PR.

v-gb committed:
tweaks (#14)

nojb commented:
> Looks good to me, modulo the PR I made against this PR.

Thanks for the PR, I checked that the manual builds OK and merged it.

Since @gasche has already approved as well, am planning to merge this PR once CI passes. Thanks a lot for your help along the way!

nojb commented:
Merged. Thanks everyone!
